### ISSUE TITLE
caseStmtMacros to support C#8-like Recursive Patterns matching (Deconstruction Pattern)

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/RFCs#57]
see https://gist.github.com/Araq/169d1e24b2b996d024a780ef6a4e6c09#file-casestmtmacros-nim for context, introduced by @Araq for a 1st version of this concept.

IT reminds me of Recursive Patterns matching (Deconstruction Pattern) in C#8, see https://www.infoq.com/articles/cs8-ranges-and-recursive-patterns

```c#
switch (employee) {
 case Employee {Name: "Bassam Alugili", Company: Company(_, _, _)} employeeTmp: ...
 case Employee(_, _, ("Stratec", _, _)) employeeTmp: ...
  case Employee(_, _, Company(_, _, _)) employeeTmp: ...
}

```

* recursive could be handled later.
* But supporting `_` placeholder is very useful, would be nice to support it

* ~~furthermore, `if` could be handled as well, it should be no different from `case` in terms of capabilities (ie introduce also `ifStmtMacros`)
analog to how `if` and `case` both handle pattern matching in C#~~ **EDIT** see reply [below](https://github.com/nim-lang/Nim/issues/8649#issuecomment-413323627); this is better handled with a regular macro `matches`
```c#
if (employee is Employee(_, _, ("Stratec", _, _)) employeeTmp) { Console.WriteLine($ "The employee: {employeeTmp.Name}!");}
```

## where?
matchers.nim

## links
* https://www.gnu.org/software/emacs/manual/html_node/elisp/Pattern-matching-case-statement.html (mentioned by @kaushalmodi)
* https://www.infoq.com/articles/cs8-ranges-and-recursive-patterns
* discussed in IRC in https://gitter.im/nim-lang/Nim?at=5b747e12937eee242321301e

## snippet copied over from @araq's post
```nim
{.experimental: "caseStmtMacros".}

import macros

macro match(n: tuple): untyped =
  result = newTree(nnkIfStmt)
  let selector = n[0]
  for i in 1 ..< n.len:
    let it = n[i]
    case it.kind
    of nnkElse, nnkElifBranch, nnkElifExpr, nnkElseExpr:
      result.add it
    of nnkOfBranch:
      for j in 0..it.len-2:
        let cond = newCall("==", selector, it[j])
        result.add newTree(nnkElifBranch, cond, it[^1])
    else:
      error "'match' cannot handle this node", it
  echo repr result

case ("foo", 78)
of ("foo", 78): echo "yes"
of ("bar", 88): echo "no"
else: discard
```

**Comment by hlaaftana** [comment]
[Fusion pattern matching](https://nim-lang.github.io/fusion/src/fusion/matching)

**Comment by konsumlamm** [comment]
> [Fusion pattern matching](https://nim-lang.github.io/fusion/src/fusion/matching) does not seem to have "custom matching" or "deconstruction", but it seems plenty sufficient for most pattern matching.

Agreed, I think this RFC has been superseded by #245. `fusion/matching` implements all of the originally proposed features (recursive matching, `_` placeholder and matching in `if` via the `matches` macro).

For "custom matching" or "deconstruction", most mechanisms can be customized (write procs to support object patterns, write an `items` iterator to support seq patterns, write a `kind` proc to support variant patterns, etc.)

See also https://github.com/nim-lang/fusion/pull/61.
